[TRTLLM-3429] feat: Overlap scheduling in C++ runtime

Funatiq opened the pull request:
Overlap execution of one iteration and preparation of next iteration.
* Add `enableTrtOverlap` switch  to `ExecutorConfig`.
* Add `enable_trt_overlap` switch to `gptManagerBenchmark`.
* Disable overlap for beam search (because of synchronous `finalize` function).
* Disable overlap for speculative decoding.

Implementation details:
* Separate in-progress and finished context requests. Only finished context requests and generation requests are processed by decoder.
* Early exit based on requests' `maxNewTokens` to avoid surplus iterations.

tensorrt-cicd commented:
[PR_Github #2458](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2458/) [ run ] triggered by Bot

Funatiq committed:
disable overlap in encoder

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
feat: invokeGatherBatch

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
feat: overlap same batch

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
chore: add enableTrtOverlap to ExecutorConfig

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
disable overlap for beam search and spec decode

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
skip overlap tests with beam search or speculative decoding

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
enable overlap in GptChunkedLongContextTests

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "A30-CPP-1, A30-CPP-2"

tensorrt-cicd commented:
[PR_Github #2460](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2460/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #2458](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2458/) [ run ] completed with state `ABORTED`

Funatiq commented:
/bot run --stage-list "A30-CPP-1"

tensorrt-cicd commented:
[PR_Github #2478](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2478/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #2478](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2478/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1778 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1778/) completed with status: 'SUCCESS'

Funatiq commented:
/bot run --stage-list "A30-CPP-1"

tensorrt-cicd commented:
[PR_Github #2505](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2505/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #2505](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2505/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1799 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1799/) completed with status: 'SUCCESS'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #3059](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3059/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3059](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3059/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #2139](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2139/) completed with status: 'FAILURE'

Funatiq committed:
feat: Enable overlap in gptManagerBenchmark

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
feat: Improve early exit

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #3080](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3080/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3080](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3080/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #2149](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2149/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #3093](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3093/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3093](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3093/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2158](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2158/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #3162](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3162/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3162](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3162/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #2199](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2199/) completed with status: 'FAILURE'

Funatiq committed:
moveFinishedContextRequestsToGeneration and skip unfinished requests in updateRequests

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #3188](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3188/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #2215](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2215/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #3196](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3196/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3196](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3196/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2220](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2220/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #3288](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3288/) [ run ] triggered by Bot

dcampora commented on cpp/tensorrt_llm/batch_manager/trtGptModel.h:
```diff
@@ -69,6 +69,22 @@ class TrtGptModel : public executor::Model
+            if (!modelConfig.getSpeculativeDecodingMode().isNone())
+            {
+                mEnableTrtOverlap = false;
+                TLLM_LOG_WARNING("TRT overlap is not supported with speculative decoding and will be disabled.");
```
Could you comment on why do these restrictions (beam search, draft) apply?

Funatiq replied:
Beam search `finalize` (aka `gatherTree`) is called in synchronously in `updateRequests` and uses the current decoder state, so the next overlap iteration would overwrite the state. So it doesn't make sense to run with overlap since we don't know when the generation is finished and finalize will run. It could be fixed by running `finalize` conditionally in decoder `forwardAsync`.

Draft-target speculative decoding only runs context phase, so it might actually work there. We can try to enable overlap separately for this. 
Other speculative decoding techniques manage more buffers and copy data from device to host and back. It would require significantly more work to get this working with overlap.

dcampora commented:
A few nitpicks here and there but nothing big.

Funatiq commented on cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.cpp:
```diff
@@ -913,6 +914,10 @@ void TrtGptModelInflightBatching::forwardAsync(RequestList const& activeRequests
 
+        // Move context requests that reached the last context chunk to the end of the vector.
+        std::partition(currRequests.contextRequests.begin(), currRequests.contextRequests.end(),
+            [](auto const& llmReq) { return !llmReq->isLastContextChunk(); });
```
To make [this](https://github.com/NVIDIA/TensorRT-LLM/pull/3625/files#diff-eb7533d7d97021d307f52f5e42e9e450a8fd66e054ba45384d448b058e1b4dcbR40) work. The idea here is get a cleaner separation between the requests that haven't finished context phase and the ones that should run through the decoder while preserving the order of requests in the batch.

I'll add this to the comment.

Funatiq committed:
refactor: Use OptionalRef for newOutputTokens tensor

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented on cpp/tensorrt_llm/batch_manager/runtimeBuffers.cpp:
```diff
@@ -453,7 +455,7 @@ void RuntimeBuffers::setFromInputs(RequestVector const& contextRequests, Request
     rnn_state_manager::RnnStateManager* rnnStateManagerPtr, PeftTable const& peftTable,
     runtime::TllmRuntime const& runtime, runtime::ModelConfig const& modelConfig,
-    runtime::WorldConfig const& worldConfig)
+    runtime::WorldConfig const& worldConfig, bool trtOverlap, TensorPtr const& newOutputTokens)
```
Done.

Funatiq commented on cpp/tensorrt_llm/batch_manager/runtimeBuffers.cpp:
```diff
@@ -905,7 +926,7 @@ std::tuple<SizeType32, RuntimeBuffers::TensorMap const&, RuntimeBuffers::TensorM
     kv_cache_manager::BaseKVCacheManager* crossKvCacheManager, rnn_state_manager::RnnStateManager* rnnStateManager,
     PeftTable const& peftTable, TllmRuntime const& runtime, ModelConfig const& modelConfig,
-    WorldConfig const& worldConfig, bool gatherGenerationLogits)
+    WorldConfig const& worldConfig, bool gatherGenerationLogits, bool trtOverlap, TensorPtr const& newOutputTokens)
```
Done.

Funatiq commented on cpp/include/tensorrt_llm/batch_manager/runtimeBuffers.h:
```diff
@@ -275,7 +275,8 @@ class RuntimeBuffers
         PeftTable const& peftTable, runtime::TllmRuntime const& runtime, runtime::ModelConfig const& modelConfig,
-        runtime::WorldConfig const& worldConfig, bool gatherGenerationLogits);
+        runtime::WorldConfig const& worldConfig, bool gatherGenerationLogits, bool trtOverlap,
+        TensorPtr const& newOutputTokens);
```
Done.

Funatiq commented on cpp/include/tensorrt_llm/batch_manager/runtimeBuffers.h:
```diff
@@ -309,7 +310,7 @@ class RuntimeBuffers
         rnn_state_manager::RnnStateManager* rnnStateManagerPtr, PeftTable const& peftTable,
         runtime::TllmRuntime const& runtime, runtime::ModelConfig const& modelConfig,
-        runtime::WorldConfig const& worldConfig);
+        runtime::WorldConfig const& worldConfig, bool trtOverlap, TensorPtr const& newOutputTokens);
```
Done.

Funatiq commented on cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.cpp:
```diff
@@ -1519,7 +1535,8 @@ void TrtGptModelInflightBatching::prepareDistGenBufferAndDecoder(RequestVector c
             *mDecoderBuffers, mKvCacheManager.get(), mCrossKvCacheManager.get(), mRnnStateManager.get(),
-            mPeftTables[mMicroBatchId], *mRuntime, mModelConfig, mWorldConfig, getGatherGenerationLogits());
+            mPeftTables[mMicroBatchId], *mRuntime, mModelConfig, mWorldConfig, getGatherGenerationLogits(),
+            isTrtOverlap(), nullptr);
```
Done.

tensorrt-cicd commented:
[PR_Github #3288](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3288/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2289](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2289/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #3325](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3325/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3298](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3298/) [ run ] completed with state `ABORTED`
[/LLM/main/L0_MergeRequest_PR pipeline #2297](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2297/) completed with status: 'FAILURE'

Funatiq committed:
feat: Add overlap scheduling support to TRTLLMDecoder

- Updated TRTLLMDecoder to accept an `enable_overlap_scheduler` parameter.
- Modified the decoder's internal logic to utilize the overlap scheduling feature.
- Adjusted the sequence lengths handling to ensure compatibility with the new scheduling approach.
- Enhanced unit tests to include cases for the overlap scheduler with the TRTLLMDecoder.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #3392](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3392/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3392](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3392/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #2376](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2376/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #3422](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3422/) [ run ] triggered by Bot

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #3423](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3423/) [ run ] triggered by Bot

Funatiq committed:
fix: allNewTokens in PP

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #3457](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3457/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #3457](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3457/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2431 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2431/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #4096](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4096/) [ run ] triggered by Bot

Funatiq commented:
/bot kill

tensorrt-cicd commented:
[PR_Github #4096](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4096/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #4101](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4101/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4101](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4101/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2918](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2918/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #4109](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4109/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4109](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4109/) [ run ] completed with state `ABORTED`

Funatiq commented:
/bot run --stage-list "A30-CPP-1, A30-CPP-2, H100_PCIe-CPP-1"

tensorrt-cicd commented:
[PR_Github #4142](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4142/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4142](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4142/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2957 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2957/) completed with status: 'SUCCESS'

Funatiq commented:
/bot run --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #4167](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4167/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4167](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4167/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2977 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2977/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --stage-list "DGX_H100-4_GPUs-CPP-1"

tensorrt-cicd commented:
[PR_Github #4184](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4184/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4184](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4184/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2991 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2991/) completed with status: 'SUCCESS'

Funatiq commented:
/bot skip --comment "All relevant tests have passed."

tensorrt-cicd commented:
[PR_Github #4209](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4209/) [ skip ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4209](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4209/) [ skip ] completed with state `SUCCESS`
Skipping testing for commit `71b8c58`
